src/main/scala/org/apache/spark/shuffle/ucx/NvkvWrapper.scala

```diff
@@ -0,0 +1,298 @@
+  packData.putInt(nvkvCtxSize)
+  packData.put(nvkvCtx)
+  packData.putLong(UnsafeUtils.getAdress(nvkvRemoteReadBuffer))
+  packData.putLong(nvkvNumOfReadBuffers*nvkvRemoteReadBufferSize)
+  packData.putInt(nvkvRemoteReadBufferSize.toInt)
+  packData.putInt(mkeyBuffer.capacity())
+  packData.put(mkeyBuffer)
+  packData.rewind()
+
+  verbosedNvkvLogDebug(s"LEO NvkvWrapper: packedNvkv nvkvCtx ${nvkvCtx} nvkvCtxSize ${nvkvCtxSize} bb ${UnsafeUtils.getAdress(nvkvRemoteReadBuffer)}")
+  verbosedNvkvLogDebug(s"LEO NvkvWrapper: packedNvkv packData capacity ${mkeyBuffer.capacity()} packData limit ${mkeyBuffer.limit()}")
+  logDebug(s"LEO NvkvWrapper: Finish init NVKV")
+
+  // TODO - store bb in array not ListBuffer
+  // To be used when fetching local blocks using DPU - no need to transfer data just bb idx
+  def createMPFromBuffer(buffer: ByteBuffer, bbSize: Int): ListBuffer[ByteBuffer] = {
+    var mpool: ListBuffer[ByteBuffer] = ListBuffer()
+
+    var pos = 0
+    while (pos < buffer.capacity()) {
+      var bb: ByteBuffer = buffer.slice()
+      bb.limit(bbSize)
+      mpool.append(bb)
+      pos += bbSize
+      buffer.position(pos)
+    }
+
+    buffer.position(0)
+    mpool
+  }
+
+  def pack: ByteBuffer = packData
+
+  def connectToRemote(add: Array[Byte]): Unit = {
+    try {
+      nvkv.connect(add)
+    } catch {
+      case e: NvkvException => logDebug("LEO NvkvWrapper: Failed to connect to remote nvkv received from DPU")
+    }
+  }
+  
+  protected class Request(private var buffer: ByteBuffer, private var length: Long, private var offset: Long) {
+    private var complete = false
+    def setComplete(): Unit = {
+      complete = true
+    }
+
+    def getComplete(): Boolean = {
+      complete
+    }
+
+    def getLength: Long = length
+    def getOffset: Long = offset
+    def getBuffer: ByteBuffer = buffer
+  }
+
+  class WriteRequest(private var dsIdx: Int, private var source: ByteBuffer, length: Long, offset: Long) extends Request(source, length, offset) {
+    def getDsIdx: Int = dsIdx
+  }
+
+  class ReadRequest(private var dest: ByteBuffer, length: Long, offset: Long) extends Request(dest, length, offset) {
+  }
+
+  private class ReducePartition(private var offset: Long, private var length: Long) {
+    def getOffset: Long = offset
+    def getLength: Long = length
+  }
+
+  private def post(request: WriteRequest) = {
+    verbosedNvkvLogDebug(s"LEO NvkvWrapper post write")
+    if (request.getOffset % 512 != 0) {
+      throw new IllegalArgumentException(s"write Illegal offset ${request.getOffset}")
+    }
+    
+    if (request.getLength % 512 != 0) {
+      throw new IllegalArgumentException(s"write Illegal length ${request.getLength}")
+    }
+
+    try nvkv.postWrite(request.getDsIdx, nvkvWriteBuffer, 0, request.getLength, request.getOffset, new Nvkv.Context.Callback() {
+      def done(): Unit = {
+        request.setComplete()
+        verbosedNvkvLogDebug(s"LEO NvkvWrapper post completed!")
+      }
+    })
+    catch {
```

Comment:
What's the point of catching & rethrowing w/o actually doing anything?

Reply:
Good point